frontend/src/metabase/parameters/components/ParameterLinkedFilters/ParameterLinkedFilters.tsx

```diff
@@ -38,27 +38,107 @@ const ParameterLinkedFilters = ({
-
-  const filteringParameters = useMemo(
-    () => parameter.filteringParameters ?? [],
-    [parameter],
-  );
-
   const usableParameters = useMemo(
     () => otherParameters.filter(usableAsLinkedFilter),
     [otherParameters],
   );
 
+  return (
+    <SectionRoot>
+      <SectionHeader>{t`Limit this filter's choices`}</SectionHeader>
+      <Content
+        usableParameters={usableParameters}
+        parameter={parameter}
+        onChangeFilteringParameters={onChangeFilteringParameters}
+        onShowAddParameterPopover={onShowAddParameterPopover}
+      />
+    </SectionRoot>
+  );
+};
+
+function Content({
+  usableParameters,
+  parameter,
+  onChangeFilteringParameters,
+  onShowAddParameterPopover,
+}: {
+  usableParameters: Parameter[];
+  parameter: Parameter;
+  onChangeFilteringParameters: (filteringParameters: ParameterId[]) => void;
+  onShowAddParameterPopover: () => void;
+}) {
+  if (usableParameters.length === 0) {
+    return (
+      <NoUsableParameters
+        onShowAddParameterPopover={onShowAddParameterPopover}
+      />
+    );
+  }
+  if (parameter.values_source_type != null) {
+    return <ParametersFromOtherSource />;
+  }
+  return (
+    <UsableParameters
+      parameter={parameter}
+      usableParameters={usableParameters}
+      onChangeFilteringParameters={onChangeFilteringParameters}
+    />
+  );
+}
+
+function NoUsableParameters({
+  onShowAddParameterPopover,
+}: {
+  onShowAddParameterPopover: () => void;
+}): JSX.Element {
+  return (
+    <div>
+      <SectionMessage>
+        {t`If you have another dashboard filter, you can limit the choices that are listed for this filter based on the selection of the other one.`}
+      </SectionMessage>
+      <SectionMessage>
+        {jt`So first, ${(
+          <SectionMessageLink key="link" onClick={onShowAddParameterPopover}>
+            {t`add another dashboard filter`}
+          </SectionMessageLink>
+        )}.`}
+      </SectionMessage>
+    </div>
+  );
+}
+
+function ParametersFromOtherSource(): JSX.Element {
+  return (
+    <div>
+      <SectionMessage>
+        {t`If the filter has values that are from another question or model, or a custom list, then this filter can't be limited by another dashboard filter.`}
+      </SectionMessage>
+    </div>
+  );
+}
+
+function UsableParameters({
+  parameter,
+  usableParameters,
+  onChangeFilteringParameters,
+}: {
+  parameter: Parameter;
+  usableParameters: Parameter[];
+  onChangeFilteringParameters: (filteringParameters: ParameterId[]) => void;
+}): JSX.Element {
+  const [expandedParameterId, setExpandedParameterId] = useState<ParameterId>();
+
   const handleFilterChange = useCallback(
     (otherParameter: Parameter, isFiltered: boolean) => {
       const newParameters = isFiltered
-        ? filteringParameters.concat(otherParameter.id)
-        : filteringParameters.filter(id => id !== otherParameter.id);
+        ? (parameter.filteringParameters ?? []).concat(otherParameter.id)
```

Comment:
That `const filteringParameters = useMemo(/* ... */)` usage was justified as per [my comment](https://github.com/metabase/metabase/pull/34358#discussion_r1347323663): the computed value was an object (array is an object, i.e. `typeof [] === 'object'`) and this object was used as a dependency of other hooks.

I'd either keep it for code clarity, or remove all hooks down the usage chain (they're all redundant: this `useCallback`, hooks in `LinkedParameter` & `Toggle`).

Reply:
I have to admit I'm confused. I've removed `const filteringParameters = useMemo(/* ... */)` because I found `handleFilterChange` can use `parameters.filteringParameters` as a dependency instead.  `handleFilterChange` doesn't need to depend on a computed object. And `LinkedParameter` can just take `!!parameter.filteringParameters?.includes(otherParameter.id)` as a parameter, instead of `filteringParameters.includes(otherParameter.id)`. There's no reference being updated on every re-render as you mention in the comment.